File: Algorithms/034_SearchforARange.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        if len(nums) == 0:
            return [-1, -1]
        low, high, middle = 0, len(nums)-1, 0
        while low <= high:
            middle = low + (high-low)/2;
            if nums[middle] < target:
                low = middle + 1
            elif nums[middle] > target:
                high = middle - 1
            if nums[middle] == target:
                break
        if nums[middle] != target:
            return [-1, -1]
        low = middle
        while low >= 0 and nums[low] == target:
            low -= 1
        high = middle
        while high < len(nums) and nums[high] == target:
            high += 1
        return  low+1, high-1

File: Algorithms/java/034.SearchforARange.java
public class Solution {
    public int[] searchRange(int[] A, int target) {
        int[] v = new int[2];
        if (A.length == 0) {
            v[0] = -1;
            v[1] = -1;
            return v;
        }
        int low = 0;
        int high = A.length-1;
        int middle = 0;
        while (low <= high) {
            middle = low + (high-low)/2;
            if (A[middle] < target) {
                low = middle + 1;
            }
            else if (A[middle] > target) {
                high = middle - 1;
            }
            if (A[middle] == target) {
                break;
            }
        }
        if (A[middle] != target) {
            v[0] = -1;
            v[1] = -1;
            return v;
        }
        low = middle;
        while (low >= 0 && A[low] == target) {
            low --;
        }
        v[0] = low+1;
        high = middle;
        while (high < A.length && A[high] == target) {
            high ++;
        }
        v[1] = high-1;
        return v;
    }
}
